helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/OptimalAssignment.java

```diff
@@ -0,0 +1,83 @@
+package org.apache.helix.controller.rebalancer.waged.model;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.helix.controller.rebalancer.waged.constraints.HardConstraint;
+import org.apache.helix.model.ResourceAssignment;
+
+/**
+ * The data model represents the optimal assignment of N replicas assigned to M instances;
+ * It's mostly used as the return parameter of an assignment calculation algorithm; If the algorithm
+ * failed to find
+ * optimal assignment given the endeavor, the user could check the failure reasons
+ */
+public class OptimalAssignment {
+  private Map<AssignableNode, List<AssignableReplica>> _optimalAssignment = new HashMap<>();
+  private Map<AssignableReplica, Map<AssignableNode, List<HardConstraint>>> _failedAssignments =
+      new HashMap<>();
+  private final ClusterContext _clusterContext;
+
+  public OptimalAssignment(ClusterContext clusterContext) {
+    _clusterContext = clusterContext;
+  }
+
+  public ClusterContext getClusterContext() {
+    return _clusterContext;
+  }
+
+  public Map<String, ResourceAssignment> getOptimalResourceAssignment() {
+    // TODO: convert the optimal assignment to map
+    return Collections.emptyMap();
+  }
+
+  public void trackAssignmentFailure(AssignableReplica replica,
+      Map<AssignableNode, List<HardConstraint>> failedAssignments) {
+    _failedAssignments.put(replica, failedAssignments);
+  }
+
+  public boolean hasAnyFailure() {
+    return !_failedAssignments.isEmpty();
+  }
+
+  public String getFailures() {
+    // TODO: format the error string
+    return _failedAssignments.toString();
+  }
+
+  public void addAssignment(AssignableReplica replica, AssignableNode node) {
```

Comment:
How you prevent the caller from sending a newly created node/replica object as the input here? I mean an object that is not in the cluster context.
If anyone makes that mistake, the following logic might generate an inconsistent result, right?

I strongly suggest you keep using the ClusterModel for now. Let's don't have these update methods for now. As we talked, we can do this refactoring later as a separate improvement.
If you want to finish this interface, you will need to add all the test cases.